app/Main.hs

```diff
@@ -1,4 +1,112 @@
+import Data.Either (partitionEithers)
+import Data.Foldable (for_, foldl')
+import Data.Map.Strict (Map)
+import Data.Maybe (catMaybes, mapMaybe, isJust)
+import Data.Version (showVersion)
+
+import Data.Map.Strict qualified as Map
+
+import Agda.Main
+import Agda.TypeChecking.Pretty (text, prettyTCM, (<+>))
+import Agda.Compiler.Backend
+import Agda.Syntax.Internal
+import Agda.Syntax.TopLevelModuleName (TopLevelModuleName)
+import Agda.Core.ToCore
+import Agda.Utils.Either (maybeRight)
+
+import Scope.Core as Scope
+import Scope.In   as Scope
+
+import Paths_agda_core
+
+
 main :: IO ()
-main = putStrLn "Hello, Agda Core"
+main = runAgda [Backend backend]
+
+
+backend :: Backend' () () () () Definition
+backend = Backend'
+  { backendName           = "agda-core"
+  , backendVersion        = Just (showVersion version)
+  , options               = ()
+  , commandLineFlags      = []
+  , isEnabled             = \ _       -> True
+  , preCompile            = \ _       -> pure ()
+  , postCompile           = \ _ _ _   -> pure ()
+  , preModule             = \ _ _ _ _ -> pure $ Recompile ()
+  , postModule            = \ _ _     -> checkModule
+  , compileDef            = \_ _ _    -> pure
+  , scopeCheckingSuffices = False
+  , mayEraseType          = \ _       -> pure True
+  }
+
+
+-- | Given a list of definitions,
+--   construct definition and constructor membership proofs, along with constructor arity.
+getModuleScope :: [Definition] -> (Defs, Cons)
+getModuleScope defs =
+  let ps = iterate Scope.inThere Scope.inHere
+      (ds, cs) :: ([QName], [QName])
+        = partitionEithers $ flip mapMaybe defs \def ->
+            let name = defName def
+            in case theDef def of
+              Datatype{}    -> Just $ Left  name
+              Function{}    -> Just $ Left  name
+              Record{}      -> Just $ Left  name
+              Constructor{} -> Just $ Right name
+              _             -> Nothing
+  in ( Map.fromList $ zip (reverse ds) ps
+     , Map.fromList $ zip (reverse cs) ps
+     )
+
+checkModule :: IsMain -> TopLevelModuleName -> [Definition] -> TCM ()
+checkModule IsMain tlm defs = do
+
+  reportSDoc "agda-core.check" 5 $ text "Checking module" <+> prettyTCM tlm
+
+  let (!gdefs, !gcons) = getModuleScope defs
+
+  reportSDoc "agda-core.check" 5 $ text "Module definitions:"  <+> prettyTCM (Map.keys gdefs)
+  reportSDoc "agda-core.check" 5 $ text "Module constructors:" <+> prettyTCM (Map.keys gcons)
+
+  for_ defs \def -> do
+
+    let Defn{defName, defType} = def
+    let dn = prettyTCM defName
+
+    reportSDoc "agda-core.check" 5 $ text "Checking" <+> dn
+
+    case convert gdefs gcons (unEl defType) of
+      Left e   -> reportSDoc "agda-core.check" 5 $
+                        text "Couldn't convert type of" <+> dn
+                    <+> text "to core syntax:" <+> text e
+      Right ty -> reportSDoc "agda-core.check" 5 $ text "Type:" <+> text (show ty)
+
+    case theDef def of
```

Comment:
Since I don't know how we're gonna interact with the type checker yet, I'd rather keep translation of terms and checking of definitions separate for now.